File: tools/creatureHero/convert.py
"""Create an Aralia master creature mesh through Microsoft's hosted TRELLIS.2.

This is the canonical authenticated Hugging Face route for hero assets because
the Python Gradio client carries HF_TOKEN into ZeroGPU jobs. It turns the staged
reference image into a high-detail master.glb. The separate optimize.mjs stage
then reduces that master to Aralia's 30,000-triangle runtime budget.

Usage: py tools/creatureHero/convert.py <entryId> [--space microsoft/TRELLIS.2] [--base directory]
"""
import json
import os
import shutil
import sys
from datetime import datetime, timezone
from pathlib import Path

from gradio_client import Client, handle_file

# ============================================================================
# Live TRELLIS export contract
# ============================================================================
# Microsoft TRELLIS.2 enforced a 100,000-face minimum during the authenticated
# 2026-08-12 pipeline proof. Exporting the larger master preserves detail, while
# optimize.mjs remains the single owner of Aralia's smaller runtime budget.
TRELLIS_EXPORT_FACE_TARGET = 100_000
TRELLIS_TEXTURE_SIZE = 1024

# ============================================================================
# Requested hero asset and local stage paths
# ============================================================================
# The command operates on one existing reference image and refuses to invent a
# replacement when the earlier collection stage has not completed.
if len(sys.argv) < 2:
    print("usage: py tools/creatureHero/convert.py <entryId> [--space owner/name]", file=sys.stderr)
    sys.exit(1)

entry_id = sys.argv[1]
space = "microsoft/TRELLIS.2"
if "--space" in sys.argv:
    space = sys.argv[sys.argv.index("--space") + 1]

# Hero Lab passes a job-specific scratch root here. The public library remains
# the command-line default so existing manual usage keeps working unchanged.
base = Path("public/creatures3d/hero")
if "--base" in sys.argv:
    base = Path(sys.argv[sys.argv.index("--base") + 1])
hero_dir = base / entry_id
reference = hero_dir / "reference.png"
master = hero_dir / "master.glb"
record_path = hero_dir / "hero.json"

if not reference.exists():
    print(f'stage "reference" artifact missing for {entry_id} — run the earlier stage first', file=sys.stderr)
    sys.exit(1)

# ============================================================================
# Authenticated remote generation
# ============================================================================
# HF_TOKEN is read from the process environment so the private credential never
# becomes part of the command arguments, generated files, or repository history.
token = os.environ.get("HF_TOKEN")
print(f"connecting to {space}… (token: {'yes' if token else 'no'})")
client = Client(space, hf_token=token)

print("start_session…")
client.predict(api_name="/start_session")

print("preprocess_image…")
processed = client.predict(input=handle_file(str(reference)), api_name="/preprocess_image")
print("preprocessed:", str(processed)[:160])

print("image_to_3d… (the slow GPU part)")
client.predict(image=handle_file(str(processed)) if isinstance(processed, str) else processed, seed=1, api_name="/image_to_3d")

print("extract_glb…")
# TRELLIS produces a reusable high-detail master here. The next local pipeline
# stage performs the game-specific triangle reduction and verifies its budget.
result = client.predict(
    decimation_target=TRELLIS_EXPORT_FACE_TARGET,
    texture_size=TRELLIS_TEXTURE_SIZE,
    api_name="/extract_glb",
)
print("extract result:", str(result)[:300])

# ============================================================================
# Downloaded GLB selection
# ============================================================================
# Gradio versions return either one cached filepath or a collection containing
# it. Accept those known shapes and fail loudly if no real GLB reached the client.
glb_path = None
candidates = result if isinstance(result, (list, tuple)) else [result]
for c in candidates:
    if isinstance(c, str) and c.lower().endswith(".glb"):
        glb_path = c
        break
    if isinstance(c, dict) and str(c.get("value", "")).lower().endswith(".glb"):
        glb_path = c["value"]
        break
if glb_path is None or not Path(glb_path).exists():
    print(f"no GLB in extract result: {result!r}", file=sys.stderr)
    sys.exit(1)

shutil.copyfile(glb_path, master)

# ============================================================================
# Durable stage provenance
# ============================================================================
# Record which hosted Space produced the master so later optimization, review,
# and approval never confuse this asset with the earlier code-sculpt candidate.
record = {"entryId": entry_id, "stages": {}, "status": "generated"}
if record_path.exists():
    record = json.loads(record_path.read_text(encoding="utf-8"))
record.setdefault("stages", {})["master"] = {
    "at": datetime.now(timezone.utc).isoformat(),
    "note": space,
}
record_path.write_text(json.dumps(record, indent=2) + "\n", encoding="utf-8")
print(f"master.glb written ({master.stat().st_size / 1024 / 1024:.1f} MB)")

File: tools/creatureHero/convert.mjs
/**
 * This compatibility wrapper creates an Aralia master creature mesh through a
 * hosted TRELLIS Space. Use convert.py for authenticated Microsoft TRELLIS.2
 * runs: the Python Gradio client was the route proven end to end on 2026-08-12,
 * while this JavaScript client did not carry the saved token into ZeroGPU.
 *
 * A successful run writes the high-detail master.glb. optimize.mjs then owns
 * the separate reduction to Aralia's 30,000-triangle runtime budget.
 *
 * Usage: npx tsx tools/creatureHero/convert.mjs <entryId> [--base dir]
 */
import { Client, handle_file } from '@gradio/client';
import { readFileSync, writeFileSync } from 'fs';
import path from 'path';
import { pathToFileURL } from 'url';

const [entryId, ...rest] = process.argv.slice(2);
if (!entryId) {
  console.error('usage: npx tsx tools/creatureHero/convert.mjs <entryId>');
  process.exit(1);
}
const baseFlag = rest.indexOf('--base');
const baseDir = baseFlag >= 0 ? rest[baseFlag + 1] : 'public/creatures3d/hero';
const spaceFlag = rest.indexOf('--space');
// Official TRELLIS.2 (higher fidelity) by default; the community TRELLIS 1
// space remains reachable via --space trellis-community/TRELLIS.
const SPACE = spaceFlag >= 0 ? rest[spaceFlag + 1] : 'microsoft/TRELLIS.2';

// Microsoft TRELLIS.2 enforced this minimum during the live 2026-08-12 proof.
// Keeping the remote export large preserves detail for the local optimizer.
const TRELLIS_EXPORT_FACE_TARGET = 100_000;
const TRELLIS_TEXTURE_SIZE = 1024;

const { assertStage, heroDir, readHero, writeHero } = await import(
  pathToFileURL(path.resolve('src/systems/entities3d/library/heroStore.ts')).href
);

assertStage(baseDir, entryId, 'reference');
const dir = heroDir(baseDir, entryId);
const referencePath = path.join(dir, 'reference.png');
const masterPath = path.join(dir, 'master.glb');

const TIMEOUT_MS = 300_000;
const deadline = setTimeout(() => {
  console.error(`TRELLIS run exceeded ${TIMEOUT_MS / 1000}s — aborting`);
  process.exit(1);
}, TIMEOUT_MS);

console.log(`connecting to ${SPACE}…`);
// A Hugging Face token (free account) lifts the anonymous ZeroGPU quota.
// Set HF_TOKEN in the environment or .env.local — never hardcode it.
const hfToken = process.env.HF_TOKEN;
if (!hfToken) console.log('no HF_TOKEN set — running on the small anonymous GPU quota');
const client = await Client.connect(SPACE, hfToken ? { hf_token: hfToken } : undefined);

// microsoft/TRELLIS.2 flow (introspected 2026-07-22 via view_api):
//   /start_session() → /preprocess_image(input) → /image_to_3d(image, seed?,
//   resolution?, …) → /extract_glb(?, decimation_target?, texture_size?)
const image = handle_file(referencePath);

async function step(name, fn) {
  try {
    return await fn();
  } catch (e) {
    console.error(`${name} failed:`);
    console.error(typeof e === 'object' ? JSON.stringify(e, Object.getOwnPropertyNames(e ?? {}), 1).slice(0, 2000) : String(e));
    process.exit(1);
  }
}

console.log('start_session…');
await step('start_session', () => client.predict('/start_session', {}));

console.log('preprocess_image…');
const pre = await step('preprocess_image', () => client.predict('/preprocess_image', { input: image }));
console.log('preprocessed:', JSON.stringify(pre.data).slice(0, 160));
// generation consumes the PREPROCESSED image (background removed by the Space)
const processed = pre.data?.[0] ?? image;

console.log('image_to_3d… (the slow GPU part)');
await step('image_to_3d', () =>
  client.predict('/image_to_3d', { image: processed, seed: 1 }),
);

console.log('extract_glb…');
// TRELLIS exports the reusable master at its supported floor. The optimizer,
// rather than the hosted service, remains the owner of the game-ready budget.
const genResult = await step('extract_glb', () =>
  client.predict('/extract_glb', {
    decimation_target: TRELLIS_EXPORT_FACE_TARGET,
    texture_size: TRELLIS_TEXTURE_SIZE,
  }),
);
const flat = JSON.stringify(genResult.data);
const urlMatch = flat.match(/"(https?:[^"]+\.glb)"/) ?? flat.match(/"url":"([^"]+\.glb)"/);
if (!urlMatch) {
  console.error('generate returned no GLB url. Raw:', flat.slice(0, 2000));
  process.exit(1);
}
const glbUrl = urlMatch[1];
console.log('downloading', glbUrl);
const res = await fetch(glbUrl);
if (!res.ok) {
  console.error(`GLB download failed: ${res.status} ${res.statusText}`);
  process.exit(1);
}
writeFileSync(masterPath, Buffer.from(await res.arrayBuffer()));

const record = readHero(baseDir, entryId) ?? { entryId, stages: {}, status: 'generated' };
record.stages.master = { at: new Date().toISOString(), note: SPACE };
writeHero(baseDir, record);
clearTimeout(deadline);
console.log(`master.glb written (${(statSizeSafe(masterPath) / 1024 / 1024).toFixed(1)} MB)`);
process.exit(0); // the gradio client keeps its socket open — leave explicitly

function statSizeSafe(p) {
  try {
    return readFileSync(p).byteLength;
  } catch {
    return 0;
  }
}
